### PR TITLE
cgal: update depends versions for 6.0.1

### PR DESCRIPTION
**Pull request by Chrismarsh**
This extends PR https://github.com/spack/spack/pull/47285 to properly include some of the required version constrains of cgal 6 including the required C++17 standard. It also adds the new no-gmp backend as a variant. The no-gmp backend seems a bit fiddly and it is unclear how it interacts with non-cmake consumers, so I have kept the gmp backend as default, matching previous cgal versions.

/cc @lang-m

### CODE REVIEW

**Comment by Chrismarsh** [comment]
The way the non-gmp backend works is a bit unclear for autotools consumers. I do wonder if this would be handled via a[setup_dependent_build_environment](https://spack.readthedocs.io/en/latest/spack.html#spack.builder.Builder.setup_dependent_build_environment) but it then has to conflict against an autotools consumer? Perhaps it is too much complexity and the gmp variant should just be removed?

**Comment by psakievich** [comment]
> The way the non-gmp backend works is a bit unclear for autotools consumers. I do wonder if this would be handled via a[setup_dependent_build_environment](https://spack.readthedocs.io/en/latest/spack.html#spack.builder.Builder.setup_dependent_build_environment) but it then has to conflict against an autotools consumer? Perhaps it is too much complexity and the gmp variant should just be removed?

I don't see a clean way to enforce this. You could make it throw with the autotools build-system if that variant is supplied.

**Comment by Chrismarsh** [comment]
Upon further reading it is not at all clear to me how the non-gmp variant should work. The variable doesn't seem to be cached (although it seems like it should be) in the installed cmake files. If a consumer wants non-gmp and are cmake, they can easily define the variable. I've opened an issue at cgal to inquire further

https://github.com/CGAL/cgal/issues/8606

For now I have left the variant code commented to save someone else the hassle, and will reopen a new PR once I have guidance. 

Boost injects cmake arguments into dependent packages 
https://github.com/spack/spack/blob/ad518d975c711c04bdc013363d8fc33a212e9194/var/spack/repos/builtin/packages/boost/package.py#L845 but I can't get this pattern to work. I've reached out to @hainest to see if they have any ideas.

For now, this PR:
- Adds the new version of cgal
- Ensures variants `shared` and `header_only` are only present on cgal < 5 as on >5 it does nothing
- Conflicts non cxx 17 compilers

File: var/spack/repos/builtin/packages/cgal/package.py
Comment (psakievich) [inline var/spack/repos/builtin/packages/cgal/package.py:76]: Why is there an upper bound on cmake? Maybe I missed it while scanning the comments?

File: var/spack/repos/builtin/packages/cgal/package.py
Comment (psakievich) [inline var/spack/repos/builtin/packages/cgal/package.py]: Is this PR ready and this is a stale comment? Or is it still WIP?

**Review by psakievich — CHANGES_REQUESTED** [review]

**Comment by psakievich** [comment]
Not clear on the status. If this isn't WIP then lots of stale comments in place that look like prototyped package implementation

**Comment by Chrismarsh** [comment]
Apologizes--  I thought I had converted it to a draft. I was waiting on the reply to https://github.com/CGAL/cgal/issues/8606 to decide how to go forward with the `+gmp` variant. 

I cannot get the boost-style `-D` consumer injection to work. I reached out to the boost maintainer @hainest as well as  @alalazo who wrote the code in the boost package, but did not hear back from either. So unless there can be a way to force consumers of this package to have a specific cmake argument injected, I don't having the variant makes sense because the consumer is, technically, who is supposed to set the GMP flag and there is nothing that is changed in the install when those flags are set.

File: var/spack/repos/builtin/packages/cgal/package.py
Comment (Chrismarsh) [reply var/spack/repos/builtin/packages/cgal/package.py:76]: I am following what is in the CMakeLists.txt
https://github.com/CGAL/cgal/blob/6.0.x-branch/CMakeLists.txt#L4

File: var/spack/repos/builtin/packages/cgal/package.py
Comment (Chrismarsh) [reply var/spack/repos/builtin/packages/cgal/package.py]: WIP sorry

**Comment by Chrismarsh** [comment]
I have removed this from the package.py but pasting it here for completness incase we need to revist this in the future. This is what I was trying, based off of boost, to inject the flags for dependent packages

```
    def setup_dependent_package(self, module, dependent_spec):
        if self.spec.satisfies("~gmp") and dependent_spec.satisfies("build_system=cmake"):
            
            args_fn = type(dependent_spec.package.builder).cmake_args
            
            def _cmake_args(self):
                return ["-DCGAL_CMAKE_EXACT_NT_BACKEND=BOOST_BACKEND"] + ["-DCGAL_DISABLE_GMP=ON"] + ["-DCMAKE_DISABLE_FIND_PACKAGE_GMP=ON"] + args_fn(self)
            
            type(dependent_spec.package.builder).cmake_args = _cmake_args

```

**Comment by Chrismarsh** [comment]
@psakievich Ok, this should be ready for review. I decided to keep the `+gmp` variant as it does impact the `depends_on` on `gmp` and `mpfr`. 

There are some stale `depends_on`, but I did not add those.

**Review by psakievich — APPROVED** [review]

**Comment by hainest** [comment]
> Upon further reading it is not at all clear to me how the non-gmp variant should work. The variable doesn't seem to be cached (although it seems like it should be) in the installed cmake files. If a consumer wants non-gmp and are cmake, they can easily define the variable. I've opened an issue at cgal to inquire further
> 
> [CGAL/cgal#8606](https://github.com/CGAL/cgal/issues/8606)
> 
> For now I have left the variant code commented to save someone else the hassle, and will reopen a new PR once I have guidance.
> 
> Boost injects cmake arguments into dependent packages
> 
> https://github.com/spack/spack/blob/ad518d975c711c04bdc013363d8fc33a212e9194/var/spack/repos/builtin/packages/boost/package.py#L845
> but I can't get this pattern to work.

I just used [the docs](https://spack.readthedocs.io/en/latest/packaging_guide.html#runtime-and-build-time-environment-variables). If that doesn't work for you, I will try to help if I can.


>I've reached out to @hainest to see if they have any ideas.

Did you reach out via email? If so, I didn't see it.

> 
> For now, this PR:
> 
>     * Adds the new version of cgal
> 
>     * Ensures variants `shared` and `header_only` are only present on cgal < 5 as on >5 it does nothing
> 
>     * Conflicts non cxx 17 compilers

**Pull request merged**
